src/app/services/conflicts.service.ts

```diff
@@ -0,0 +1,128 @@
+   */
+  public populateConflictsCache(data) {
+    /* NOTE: The data in the conflicts cache MUST be sorted by section ID.
+     * Currently we are assuming the conflicts in the API will always be sorted.
+     * If that is not the case, a sort must be implemented here.
+     */
+    var courselen = data.courses.length;
+    for (var i = 0; i < courselen; ++i) {
+      var sectlen = data.courses[i].sections.length;
+      for (var j = 0; j < sectlen; ++j) {
+        var id = data.courses[i].sections[j].id;
+
+        // TODO some method of cache expiration
+        if (!(id in this.conflicts)) {
+          this.conflicts[id] = data.courses[i].sections[j].conflicts;
+        }
+      }
+    }
+  };
+
+  /**
+   * Flatten the selection data into an array of arrays [sectionId, courseId]
+   * sorted by section id. Also generate a map of course ids to the number of
+   * sections that are selected by that course.
+   * This is a preprocessing step for evaluating conflicts on each section
+   * efficiently (in O(k + s) time) without having to refer back to the selections
+   * in their normal format.
+   * Return an object containing the flattened array and the course id map.
+   * @param {Object} selections - The selections object returned verbatim from Yacs.user.getSelections().
+   * @return {Object} The selection data converted into an array of 2-element arrays, each with a section id and its associated course id, and a map of each course ID to its number of selected sections.
+   */
+  public flattenSelections(selections) {
+    var selectionsFlat = [];
+    var courseSelectionCounts = {};
+
+    for (var cid in selections) {
+      courseSelectionCounts[cid] = 0;
+      var sidlen = selections[cid].length;
+
+      for (var i = 0; i < sidlen; ++i) {
+        var sid = selections[cid][i];
+        selectionsFlat.push([sid, cid]);
+        courseSelectionCounts[cid]++;
+      }
+    }
+
+    // then sort selectionsFlat by ascending order of sid
+    // this could be improved with an optimized k-merge of the sectionId arrays, TODO
+    selectionsFlat.sort(function(sidcid1, sidcid2) {
+      // the arguments are those 2-element arrays [sid, cid] from above
+      return sidcid1[0] - sidcid2[0];
+    });
+    return {
+      'selectionsFlat': selectionsFlat,
+      'courseSelectionCounts': courseSelectionCounts
+    };
+  }
+
+  /** Given a section ID, the output of flattenSelections,
+   * (and data in the conflicts cache),
+   * return a boolean of whether this section has any conflicts with
+   * current selections.
+   * @param {int} sectId - Section ID to check for conflicts.
+   * @return {boolean} Whether this section conflicts with currently selected sections.
+   */
+  public doesConflict(sectId: number) {
+    let flattenedSelections = this.flattenSelections(JSON.parse(localStorage.getItem('selections'))); // An object in the same format returned by flattenSelections.
```

Comment:
We don't want to depend on the implementation of selections here. We should inject the selections service and get selections that way.